src/dialogs/sendto/Services/Manager.vala

```diff
@@ -0,0 +1,206 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright Budgie Desktop Developers, elementary LLC
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+public class Bluetooth.ObjectManager : Object {
+	public bool has_object { get; private set; default = false; }
+
+	private DBusObjectManagerClient object_manager;
+
+	public signal void device_added(Bluetooth.Device device);
+	public signal void device_removed(Bluetooth.Device device);
+	public signal void status_discovering();
+
+	construct {
+		create_manager.begin();
+		register_obex_agentmanager();
+	}
+
+	public async void create_manager() {
+		try {
+			object_manager = yield new DBusObjectManagerClient.for_bus.begin(
+				BusType.SYSTEM,
+				DBusObjectManagerClientFlags.NONE,
+				"org.bluez",
+				"/",
+				object_manager_proxy_get_type,
+				null
+			);
+
+			object_manager.get_objects().foreach((object) => {
+				object.get_interfaces().foreach((iface) => on_interface_added(object, iface));
+			});
+
+			object_manager.interface_added.connect(on_interface_added);
+
+			object_manager.interface_removed.connect(on_interface_removed);
+
+			object_manager.object_added.connect((object) => {
+				object.get_interfaces().foreach((iface) => on_interface_added(object, iface));
+			});
+
+			object_manager.object_removed.connect((object) => {
+				object.get_interfaces().foreach((iface) => on_interface_removed(object, iface));
+			});
+		} catch (Error e) {
+			critical("Error getting Bluez object manager: %s", e.message);
+		}
+	}
+
+	// TODO: Do not rely on this when it is possible to do it natively in Vala
+	[CCode (cname="bluetooth_device_proxy_get_type")]
+	extern static GLib.Type get_device_proxy_type();
+
+	[CCode (cname="bluetooth_adapter_proxy_get_type")]
+	extern static GLib.Type get_adapter_proxy_type();
+
+	private GLib.Type object_manager_proxy_get_type(DBusObjectManagerClient manager, string object_path, string? interface_name) {
+		if (interface_name == null) return typeof (DBusObjectProxy);
+
+		switch (interface_name) {
+			case "org.bluez.Device1":
+				return get_device_proxy_type();
+			case "org.bluez.Adapter1":
+				return get_adapter_proxy_type();
+			default:
+				return typeof(DBusProxy);
+		}
+	}
+
+	private void register_obex_agentmanager() {
+		try {
+			var connection = GLib.Bus.get_sync(BusType.SESSION);
+			connection.call.begin(
+				"org.bluez.obex",
+				"/org/bluez/obex",
+				"org.bluez.obex.AgentManager1",
+				"RegisterAgent", // TODO: Do we need to worry about unregistering?
```

Comment:
Does this still need to be done? Is it handled gracefully across panel ~~crashes~~ restarts? That's probably the time I could see it unregistering